### PR TITLE
Quotation and guaranteed stop percentage from context. `applyCreateOrder` sort-out.

### PR DESCRIPTION
**Pull request by arealclimber**
### DEVELOP

- [x]   TradeTab
- [x]  TradeTabMobile
- [x]  PositionOpenModal
- [x]  PositionClosedModal
- [x]  PositionUpdatedModal
- [x] marketContext

### CHECK LIST

- [Naming convention](https://github.com/CAFECA-IO/WorkGuidelines/blob/main/technology/coding-convention/naming-convention.md) verification: checked
- Coding style verification: checked
- new Library: 0
- new Class / Component: 0
- new loop: 0
- new recursive: 0
- high risk: 0
- new sql: 0

### UML

- [Sequence Diagram 001 (interaction between user, frontend and backend)](https://github.com/CAFECA-IO/Documents/blob/main/TBD/TBDSD00001.md)
- [Sequence Diagram 002 (background mechanism)](https://github.com/CAFECA-IO/Documents/blob/main/TBD/TBDSD00002.md)
- [Component Diagram](https://github.com/CAFECA-IO/Documents/blob/main/TBD/TBDCP00001.md)
- [Class Diagram](https://github.com/CAFECA-IO/Documents/blob/main/TBD/TBDCD00001.md)
- [Global context document](https://www.notion.so/d93ac7c941c94fb6b173d34a7bc4e959)
- [Table of interface](https://github.com/CAFECA-IO/EventReport/blob/main/meeting/20230106_tbd_mvp.md)

### CODE REVIEW

File: src/components/trade_tab/trade_tab.tsx [hunk 1]
```diff
@@ -103,21 +103,192 @@ const TradeTab = () => {
...
+
+      if (longQuotation.success && long.typeOfPosition === TypeOfPosition.BUY) {
+        setLongQuotation(long);
+
+        // Deprecated: before merging into develop (20230327 - Shirley)
+        // eslint-disable-next-line no-console
+        // console.log('long ref in effect', longQuotationRef.current);
+      }
+
+      // ToDo: handle the error code (20230327 - Shirley)
```
Comment (arealclimber) [inline src/components/trade_tab/trade_tab.tsx:246]: - 拿不到報價，按鈕反灰
- 告知用戶報價異常

File: src/components/trade_tab/trade_tab.tsx [hunk 2]
```diff
@@ -103,21 +103,192 @@ const TradeTab = () => {
...
+        // console.log('INVAILD_INPUTS');
+      } else if (longQuotation.code === Code.SERVICE_TERM_DISABLE) {
+        // console.log('SERVICE_TERM_DISABLE');
+      } else if (longQuotation.code === Code.INTERNAL_SERVER_ERROR) {
+        // console.log('INTERNAL_SERVER_ERROR');
+      }
+    } catch (err) {
+      // ToDo: handle the error code (20230327 - Shirley)
+      // console.error(err);
+    }
```
Comment (arealclimber) [inline src/components/trade_tab/trade_tab.tsx:259]: - market context 壞掉，拿不到報價

File: src/components/trade_tab_mobile/trade_tab_mobile.tsx [hunk 3]
```diff
@@ -201,40 +372,86 @@ const TradeTabMobile = () => {
...
+        asset: unitAsset,
+        amount: requiredMarginLongRef.current,
+      },
+      quotation: {
+        ticker: marketCtx.selectedTicker?.currency ?? '',
+        typeOfPosition: TypeOfPosition.SELL,
+        targetAsset: marketCtx.selectedTicker?.currency ?? '',
+        unitAsset: unitAsset,
+        price: Number(sellPrice) ?? 9999999999,
+        deadline: getTimestamp() + POSITION_PRICE_RENEWAL_INTERVAL_SECONDS,
```
Comment (arealclimber) [inline src/components/trade_tab_mobile/trade_tab_mobile.tsx:425]: - default price 取代 9999999999

**Comment by arealclimber** [comment]
Update form modal 的 GSL 沒有更新

File: src/components/position_closed_modal/position_closed_modal.tsx [hunk 4]
```diff
@@ -204,6 +223,39 @@ const PositionClosedModal = ({
...
+      ) {
+        return data;
+      } else {
+        // Deprecated: before merging into develop (20230327 - Shirley)
+        // eslint-disable-next-line no-console
+        console.log('diable submit');
+        setQuotationError(true);
+      }
+    } catch (err) {
+      setQuotationError(true);
```
Comment (TzuHanLiang) [inline src/components/position_closed_modal/position_closed_modal.tsx:255]: getCFDQuotation 的 error handle 有在 context 裡面做了，這邊應該用 result.success 判斷就可以了， result.success === true 的話是成功 result.success === false 就是失敗，不過也可以有

File: src/contexts/market_context.tsx [hunk 5]
```diff
@@ -264,7 +264,7 @@ export const MarketProvider = ({children}: IMarketProvider) => {
     } catch (error) {
       result = defaultResultFailed;
     }
-    return result;
+    return {...result}; // Info: (20230328 - Shirley) return a new object to avoid the reference of result.data being changed
```
Comment (TzuHanLiang) [inline src/contexts/market_context.tsx:267]: 應該不需要這樣調整的

**Review by TzuHanLiang — COMMENTED** [review]
- marketContext 我這邊認為不需要調整

File: src/contexts/market_context.tsx (hunk 5 above)
Comment (Luphia) [reply src/contexts/market_context.tsx:267]: - call by reference issue
```ts
const def = { result: false, data: Math.random() }
const f = () => {
    const result = def;
    result.data = Math.random();
    return result;
}
const rs1 = f();
const rs2 = f()
// rs1.data always equals rs2.data
```

Comment (Luphia) [reply src/contexts/market_context.tsx:267]: - by the way 這並未解決根源問題，需可參考 [object clone](https://stackoverflow.com/questions/28150967/typescript-cloning-object) 方法

File: src/components/position_closed_modal/position_closed_modal.tsx (hunk 4 above)
Comment (Luphia) [reply src/components/position_closed_modal/position_closed_modal.tsx:255]: - 還存在 getCFDQuotation 回傳資料不符需求、資料格式轉換為 IQuotation 失敗、其他未知錯誤可能

**Review by Luphia — APPROVED** [review]
- no bug found

**Pull request merged**